treelog/iter.py

```diff
@@ -1,58 +1,61 @@
-import itertools, functools, warnings, inspect
+import itertools, functools, warnings, inspect, typing, types
+from . import proto
 
-class wrap:
+T = typing.TypeVar('T')
+
+class wrap(typing.Generic[T]):
   '''Wrap iterable in consecutive title contexts.
 
   The wrapped iterable is identical to the original, except that prior to every
   next item a new log context is opened taken from the ``titles`` iterable. The
   wrapped object should be entered before use in order to ensure that this
   context is properly closed in case the iterator is prematurely abandoned.'''
 
-  def __init__(self, titles, iterable):
+  def __init__(self, titles: typing.Union[typing.Iterable[str], typing.Generator[str, T, None]], iterable: typing.Iterable[T]) -> None:
     self._titles = iter(titles)
     self._iterable = iter(iterable)
-    self._log = None
+    self._log = True # type: typing.Union[proto.Log, bool]
```

Comment:
why not just typing.Optional and keep _log at None?

Reply:
In the old code `self._log` could be `None` (not yet entered), a `Log` instance (entered) or `False` (exited). Replacing `None` with `True` reduces the number of types to `proto.Log` and `bool` and arguably simplifies the tests in `__iter__` and `__exit__`.

Reply:
Done.